### ISSUE TITLE
NLP

### ISSUE THREAD

**Issue by lolzbreak001-cell** [rio0hayoo/Crowd-Report-Online#6]
ChatBot NLP
# chatbot_pelaporan.py
import numpy as np
import pandas as pd
import re
import datetime
import json
import pickle
import os
from pathlib import Path

import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# NLTK 
try:
    nltk.download('punkt')
    nltk.download('stopwords')
    print("NLTK resources downloaded successfully!")
except Exception:
    print("Error downloading NLTK resources. Please check your internet connection.")

from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize, sent_tokenize

try:
    from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
    sastrawi = StemmerFactory().create_stemmer()
except Exception:
    sastrawi = None

# Tokenizer Lanjutan
class AdvancedTokenizer:
    def __init__(self):
        self.ps = PorterStemmer()  # gunakan Sastrawi jika tersedia (lihat komentar)
        try:
            self.stop_words = set(stopwords.words('indonesian'))
        except Exception:
            print("Indonesian stopwords not available. Using English stopwords instead.")
            self.stop_words = set(stopwords.words('english'))

        # Stopword kustom ID
        self.custom_stopwords = {
            'yang', 'di', 'dan', 'itu', 'dengan', 'untuk', 'pada', 'ke', 'dari',
            'dalam', 'ini', 'saya', 'kamu', 'kami', 'kita', 'mereka', 'ada', 'adalah',
            'telah', 'akan', 'atau', 'juga', 'bahwa', 'oleh', 'karena', 'jika'
        }
        self.stop_words.update(self.custom_stopwords)

    def clean_text(self, text):
        text = text.lower()
        text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)
        text = re.sub(r'@\w+|#\w+', '', text)
        text = re.sub(r'[^\w\s.]', '', text)
        text = re.sub(r'\s+', ' ', text).strip()
        return text

    def sentence_tokenize(self, text):
        try:
            return sent_tokenize(text)
        except Exception:
            return [s.strip() for s in text.split('.') if s.strip()]

    def _stem(self, token: str) -> str:
        if sastrawi is not None:
            return sastrawi.stem(token)
        return self.ps.stem(token)

    def word_tokenize(self, text, remove_stopwords=True, stemming=True):
        text = self.clean_text(text)
        try:
            tokens = word_tokenize(text)
        except Exception:
            tokens = text.split()

        if remove_stopwords:
            tokens = [t for t in tokens if t not in self.stop_words]

        if stemming:
            tokens = [self._stem(t) for t in tokens]

        return tokens

    def ngram_tokenize(self, text, n=2):
        tokens = self.word_tokenize(text, remove_stopwords=True, stemming=True)
        return [' '.join(tokens[i:i+n]) for i in range(len(tokens)-n+1)]

    def analyze_text(self, text):
        return {
            'original_text': text,
            'cleaned_text': self.clean_text(text),
            'sentences': self.sentence_tokenize(text),
            'words': self.word_tokenize(text, remove_stopwords=False, stemming=False),
            'words_processed': self.word_tokenize(text, remove_stopwords=True, stemming=True),
            'bigrams': self.ngram_tokenize(text, n=2),
            'trigrams': self.ngram_tokenize(text, n=3),
            'word_count': len(self.word_tokenize(text, remove_stopwords=False, stemming=False)),
            'unique_words': len(set(self.word_tokenize(text, remove_stopwords=True, stemming=True))),
        }

# =========================
# Util Training dari JSON
# =========================
def load_intents_json(json_path: str):
    data = json.loads(Path(json_path).read_text(encoding="utf-8"))
    X, y = [], []
    for it in data["intents"]:
        label = it["name"]
        for ex in it["examples"]:
            X.append(ex)
            y.append(label)
    return X, y

def train_intent_model(json_path: str, tokenizer: AdvancedTokenizer, out_dir="model"):
    Path(out_dir).mkdir(exist_ok=True)

    vec = TfidfVectorizer(
        tokenizer=lambda s: tokenizer.word_tokenize(s, remove_stopwords=True, stemming=True),
        ngram_range=(1, 2),
        max_features=5000
    )

    X, y = load_intents_json(json_path)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )

    Xtr = vec.fit_transform(X_train)
    Xte = vec.transform(X_test)

    clf = LogisticRegression(max_iter=2000, class_weight="balanced")
    clf.fit(Xtr, y_train)

    y_pred = clf.predict(Xte)
    print("\n=== Classification Report ===")
    print(classification_report(y_test, y_pred, digits=3))

    pickle.dump(vec, open(Path(out_dir) / "vectorizer.pkl", "wb"))
    pickle.dump(clf, open(Path(out_dir) / "model.pkl", "wb"))
    print(f"[OK] Model disimpan ke folder: {out_dir}")

# =========================
# Chatbot
# =========================
class NLPChatbot:
    def __init__(self):
        self.tokenizer = AdvancedTokenizer()
        self.ps = PorterStemmer()
        self.stop_words = self.tokenizer.stop_words

        # Pola regex bawaan
        self.patterns = {
            'greeting': [
                r'halo|hai|hi|hei|helo',
                r'selamat (pagi|siang|sore|malam)',
                r'apa kabar'
            ],
            'report_incident': [
                r'lapor|laporkan|melaporkan|pelaporan|melapor',
                r'kejadian|insiden|kecelakaan|kejahatan|pembegalan|begal|pencurian',
                r'(melihat|menyaksikan) (pencurian|perampokan|kecelakaan|penganiayaan|pembegalan)'
            ],
            'location': [
                r'(di|dekat|sekitar) (jalan|jl|rt|rw|kelurahan|kecamatan)',
                r'alamat|lokasi|tempat'
            ],
            'thanks': [
                r'terima kasih|terimakasih|makasih|thanks|thank you'
            ],
            'goodbye': [
                r'selamat tinggal|dadah|sampai jumpa|bye|goodbye'
            ],
            'help': [
                r'bantuan|help|tolong|caranya|bagaimana'
            ]
        }

        # Respons bawaan (bisa Anda tambah sesuai label dataset)
        self.responses = {
            'greeting': [
                "Halo! Saya adalah chatbot pelaporan insiden publik. Ada yang bisa saya bantu?",
                "Hai! Siap membantu pelaporan insiden publik Anda.",
                "Selamat datang! Silakan laporkan insiden yang Anda alami atau saksikan."
            ],
            'report_incident': [
                "Saya siap membantu Anda melaporkan insiden. Bisakah Anda ceritakan apa yang terjadi?",
                "Mari kita laporkan insiden tersebut. Silakan jelaskan kejadiannya secara detail.",
                "Saya memahami Anda ingin melaporkan suatu insiden. Tolong beri tahu saya apa yang terjadi."
            ],
            'location': [
                "Di mana lokasi kejadiannya?",
                "Bisakah Anda memberikan alamat atau lokasi kejadian?",
                "Di mana insiden ini terjadi?"
            ],
            'thanks': [
                "Sama-sama! Saya senang bisa membantu.",
                "Terima kasih kembali. Laporan Anda sangat berharga untuk keamanan publik.",
                "Dengan senang hati. Semoga masalah dapat segera ditangani."
            ],
            'goodbye': [
                "Selamat tinggal! Jaga keselamatan Anda.",
                "Sampai jumpa! Terima kasih telah melaporkan insiden.",
                "Dadah! Hubungi kami jika ada insiden lainnya."
            ],
            'help': [
                "Saya bisa membantu Anda melaporkan insiden seperti pencurian, kecelakaan, atau kejahatan.",
                "Ketik 'lapor' untuk mulai melaporkan insiden. Saya akan memandu Anda langkah demi langkah.",
                "Untuk melaporkan insiden, cukup ketik 'laporkan insiden' atau ceritakan apa yang terjadi."
            ],
            'default': [
                "Maaf, saya tidak mengerti. Bisakah Anda mengulangi?",
                "Saya belum memahami pertanyaan Anda. Bisa dijelaskan lebih lanjut?",
                "Bisa Anda uraikan dengan cara yang berbeda?"
            ],
            # Tambahan opsional agar selaras label dataset besar:
            'status_check': ["Baik, mohon berikan ID laporan jika ada. Saya akan cek statusnya."],
            'cancel_report': ["Baik, saya akan menandai laporan untuk ditinjau pembatalannya."],
            'evidence_info': ["Terima kasih. Laporan Anda akan diproses. Apakah ada hal lain yang ingin ditambahkan?"],
            'escalation_request': ["Jika darurat, segera hubungi 112/110. Saya catat prioritasnya."]
        }

        # Penyimpanan laporan sementara
        self.reports = []

        # Vectorizer khusus SIMILARITY (terpisah dari model training)
        self.sim_vectorizer = TfidfVectorizer(
            tokenizer=self.tokenizer.word_tokenize, max_features=1000
        )
        self.pattern_corpus = self._build_pattern_corpus()

        # === Muat model terlatih (jika ada)
        self.clf = None
        self.vec = None
        self.conf_threshold = 0.45  # tuning sesuai evaluasi Anda
        try:
            if os.path.exists("model/vectorizer.pkl") and os.path.exists("model/model.pkl"):
                self.vec = pickle.load(open("model/vectorizer.pkl", "rb"))
                self.clf = pickle.load(open("model/model.pkl", "rb"))
                print("[INFO] Intent model loaded.")
        except Exception as e:
            print("[WARN] Gagal memuat model intent:", e)

    def _build_pattern_corpus(self):
        corpus = []
        for intent, patterns in self.patterns.items():
            corpus.extend(patterns)
        return corpus

    def preprocess_text(self, text):
        return ' '.join(self.tokenizer.word_tokenize(text, remove_stopwords=True, stemming=True))

    def _map_label_to_response_key(self, label):
        mapping = {
            "greeting": "greeting",
            "help": "help",
            "report_incident": "report_incident",
            "thanks": "thanks",
            "goodbye": "goodbye",
            "location_provide": "location",
            "time_provide": "report_incident",
            "describe_detail": "report_incident",
            "victim_info": "report_incident",
            "evidence_info": "evidence_info",
            "status_check": "status_check",
            "cancel_report": "cancel_report",
            "escalation_request": "escalation_request",
            "out_of_scope": "default",
        }
        return mapping.get(label, "default")

    # ============ INTENT ============
    def classify_intent_similarity(self, text):
        processed_text = self.preprocess_text(text)
        temp_corpus = self.pattern_corpus + [processed_text]
        try:
            tfidf_matrix = self.sim_vectorizer.fit_transform(temp_corpus)
            similarities = cosine_similarity(tfidf_matrix[-1:], tfidf_matrix[:-1])
            max_similarity = float(np.max(similarities))
            max_index = int(np.argmax(similarities))
            if max_similarity > 0.3:
                index = 0
                for intent, patterns in self.patterns.items():
                    if max_index < index + len(patterns):
                        return intent
                    index += len(patterns)
        except Exception as e:
            print(f"Error in similarity matching: {e}")
        return 'default'

    def classify_intent(self, text):
        # 1) Model terlatih
        if self.clf is not None and self.vec is not None:
            try:
                v = self.vec.transform([text])
                if hasattr(self.clf, "predict_proba"):
                    proba = self.clf.predict_proba(v)[0]
                    pred = self.clf.classes_[int(np.argmax(proba))]
                    conf = float(np.max(proba))
                else:
                    pred = self.clf.predict(v)[0]
                    conf = 1.0
                if conf >= self.conf_threshold:
                    return self._map_label_to_response_key(pred)
            except Exception as e:
                print("Model inference error:", e)

        # 2) Regex
        processed_text = self.tokenizer.clean_text(text)
        for intent, patterns in self.patterns.items():
            for pattern in patterns:
                if re.search(pattern, processed_text):
                    return intent

        # 3) Similarity
        return self.classify_intent_similarity(text)

    def get_response(self, intent):
        # Jika intent belum ada di responses, jatuh ke default
        if intent not in self.responses:
            intent = 'default'
        return np.random.choice(self.responses[intent])

    # ============ Laporan ============
    def generate_report_id(self, user_id):
        timestamp = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        return f"RPT{user_id}{timestamp}"

    def process_report(self, user_id, report_text):
        report_id = self.generate_report_id(user_id)
        text_analysis = self.tokenizer.analyze_text(report_text)
        report_data = {
            'report_id': report_id,
            'user_id': user_id,
            'timestamp': str(datetime.datetime.now()),
            'report_text': report_text,
            'text_analysis': text_analysis,
            'status': 'pending',
            'word_count': text_analysis['word_count'],
            'key_phrases': text_analysis['bigrams'][:5]
        }
        self.reports.append(report_data)
        return report_id, report_data

    def show_text_analysis(self, text):
        analysis = self.tokenizer.analyze_text(text)
        print("\n" + "="*50)
        print("ANALISIS TEKS")
        print("="*50)
        print(f"Teks asli: {analysis['original_text']}")
        print(f"Teks bersih: {analysis['cleaned_text']}")
        print(f"Jumlah kalimat: {len(analysis['sentences'])}")
        print(f"Jumlah kata: {analysis['word_count']}")
        print(f"Kata unik: {analysis['unique_words']}")
        print(f"Kata-kata penting: {', '.join(analysis['words_processed'][:10])}")
        print(f"Bigrams kunci: {', '.join(analysis['bigrams'][:5])}")
        print("="*50 + "\n")

    # ============ Loop Chat ============
    def chat(self):
        print("=" * 60)
        print("CHATBOT PELAPORAN INSIDEN PUBLIK")
        print("Berbasis NLP dengan Tokenizer Advanced + Model Intent (opsional)")
        print("=" * 60)
        print("Chatbot: Halo! Saya adalah chatbot pelaporan insiden publik apakah ada yang bisa dibantu.")
        print("Chatbot: Untuk mulai melaporkan, ketik 'lapor' atau 'laporkan insiden'.")
        print("Chatbot: Ketik 'analisis' untuk melihat analisis teks.")
        print("Chatbot: Ketik 'keluar' untuk mengakhiri percakapan.")
        print("-" * 60)

        user_id = ("ANON")
        reporting_mode = False
        report_data = {}

        while True:
            user_input = input("Anda: ").strip()

            if user_input.lower() == 'keluar':
                print("Chatbot: Terima kasih telah menggunakan layanan pelaporan insiden. Sampai jumpa!")
                break

            if user_input.lower() == 'analisis':
                text_to_analyze = input("Chatbot: Masukkan teks yang ingin dianalisis: ")
                self.show_text_analysis(text_to_analyze)
                continue

            # Mode pelaporan terstruktur
            if reporting_mode:
                if 'lokasi' not in report_data:
                    report_data['lokasi'] = user_input
                    print("Chatbot: Terima kasih. Kapan insiden ini terjadi? (contoh: hari ini pukul 10.00, kemarin sore)")
                    continue

                if 'waktu' not in report_data:
                    report_data['waktu'] = user_input
                    print("Chatbot: Bisakah Anda jelaskan secara detail apa yang terjadi?")
                    continue

                if 'detail' not in report_data:
                    report_data['detail'] = user_input
                    print("Chatbot: Apakah ada korban atau kerugian? Jika ada, tolong jelaskan.")
                    continue

                if 'korban' not in report_data:
                    report_data['korban'] = user_input
                    print("Chatbot: Apakah Anda memiliki informasi pelaku atau bukti lainnya?")
                    continue

                if 'bukti' not in report_data:
                    report_data['bukti'] = user_input
                    report_text = (
                        f"Laporan insiden: {report_data['detail']}. "
                        f"Lokasi: {report_data['lokasi']}. "
                        f"Waktu: {report_data['waktu']}. "
                        f"Korban: {report_data['korban']}. "
                        f"Bukti: {report_data['bukti']}."
                    )
                    report_id, saved_report = self.process_report(user_id, report_text)

                    print("=" * 60)
                    print("Chatbot: TERIMA KASIH ATAS LAPORAN ANDA!")
                    print(f"Chatbot: Laporan Anda telah direkam dengan ID: {report_id}")
                    print(f"Chatbot: Jumlah kata dalam laporan: {saved_report['word_count']}")
                    print(f"Chatbot: Frasa kunci: {', '.join(saved_report['key_phrases'])}")
                    print(f"Chatbot: Laporan telah disimpan pada: {saved_report['timestamp']}")
                    print("Chatbot: Laporan Anda akan segera diproses oleh pihak berwenang.")
                    print("=" * 60)

                    reporting_mode = False
                    report_data = {}
                    continue

            # Klasifikasi intent
            intent = self.classify_intent(user_input)

            # Jika intent pelaporan, mulai mode pelaporan
            if intent in ('report_incident',) and not reporting_mode:
                reporting_mode = True
                report_data = {}
                print("Chatbot: Mari kita laporkan insiden tersebut. Di mana lokasi kejadiannya?")
                continue

            # Balas sesuai intent
            response = self.get_response(intent)
            print(f"Chatbot: {response}")

# =========================
# Entry point
# =========================
if __name__ == "__main__":
    import sys
    # Mode:
    # 1) Train: python chatbot_pelaporan.py train intents_id_large.json
    # 2) Chat : python chatbot_pelaporan.py
    if len(sys.argv) >= 2 and sys.argv[1].lower() == "train":
        json_path = sys.argv[2] if len(sys.argv) >= 3 else "intents_id_large.json"
        print(f"[INFO] Melatih model dari: {json_path}")
        tok = AdvancedTokenizer()
        train_intent_model(json_path, tok, out_dir="model")
    else:
        print("Mempersiapkan chatbot dengan tokenizer advanced...")
        chatbot = NLPChatbot()
        chatbot.chat()

Dataset (JSON)
[intents_id_large.json](https://github.com/user-attachments/files/22699396/intents_id_large.json)
